### PR TITLE
Add support for parsing certification requests

### PR DESCRIPTION
**Pull request by djc**
I'd like to add CSR parsing code to this crate. I'm happy to incorporate any feedback you have.

Depends on https://github.com/rusticata/oid-registry/pull/1 (via a Git dependency for now).

### CODE REVIEW

File: src/x509_parser.rs [hunk 1]
```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
...
+    for attr in v.into_iter() {
+        if attributes.insert(attr.oid.clone(), attr).is_some() {
+            // duplicate attributes are not allowed
+            return Err(Err::Failure(X509Error::DuplicateAttributes));
+        }
+    }
+    Ok((i, attributes))
+}
+
+fn parse_cri_attributes(i: &[u8]) -> X509Result<HashMap<Oid, X509CriAttribute>> {
```
Comment (chifflier) [inline src/x509_parser.rs:568]: My test with an openssl CSR failed at `parse_cri_attributes`
The created CSR has no attribute, so input is `a0 00` (empty [0] SET OF Attribute)
Adding a test for empty input solved the problem

File: src/x509_parser.rs [hunk 2]
```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
...
+    let (i, hdr) = ber_read_element_header(i).or(Err(Err::Error(X509Error::InvalidAttributes)))?;
```
Comment (chifflier) [inline src/x509_parser.rs:569]: I think using `parse_ber_tagged_implicit_g` would have been better, as it would ensure only the content of the sequence is parsed.
I will do the update

File: src/x509_parser.rs [hunk 3]
```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
...
+    (0..hdr.structured)
+        .into_iter()
```
Comment (chifflier) [inline src/x509_parser.rs:571]: If I get that correctly, this is equivalent to a `if hdr.structured != 0` statement ?

File: src/x509_parser.rs [hunk 4]
```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
...
+            subject,
+            subject_pki,
+            attributes,
+            raw: &start_i[..len],
+        };
+        Ok((i, tbs))
+    })(i)
+}
+
+pub fn parse_x509_csr_der<'a>(i: &'a [u8]) -> X509Result<X509CertificationRequest<'a>> {
```
Comment (chifflier) [inline src/x509_parser.rs:618]: This function is lacking doc (and tests)

**Review by chifflier — APPROVED** [review]
Hi,
Thank you for your contribution! I have a few comments in the PR, but at that point you do not need to change your commits or resubmit a PR.
I'm running some tests, and will add additional patches if needed. I will merge it soon.

File: src/x509_parser.rs (hunk 3 above)
Comment (djc) [reply src/x509_parser.rs:571]: I guess it might be, I probably don't have the same level of ASN.1 understanding you do. :)

Comment (djc) [reply src/x509_parser.rs:571]: Also note that I have only tested this on CSRs that have exactly one attribute, so this could perhaps benefit from testing with 0 and more-than-1 attributes (though I'm not clear if any other attributes are commonly in use).

File: src/x509_parser.rs (hunk 4 above)
Comment (djc) [reply src/x509_parser.rs:618]: If you focus on the other stuff, shall I focus on adding some docs and tests?

Comment (djc) [reply src/x509_parser.rs:618]: (Or should I wait on seeing what tests you've already written?)

**Comment by chifflier** [comment]
Merged, with some changes to make peace with clippy
I added one test with a CSR, and fixed the issue with empty attributes list.
Any change or addition can be done using new PRs.
Thanks!
